effekt/shared/src/main/scala/effekt/core/Recursive.scala

```diff
@@ -0,0 +1,132 @@
+package effekt.core
+
+import scala.collection.mutable
+
+// TODO: Is there a simpler way of iterating the functions and calls (like a fold)?
+//       Reachable etc. use similar recursions
+//       We could probably remove a lot of recursion here!
+
+/**
+ * Gather all functions, their arguments, and arguments of their recursive calls
+ */
+
+ case class FunctionGathering(
+    cparams: List[Id],
+    vparams: List[Id],
+    bparams: List[Id],
+    cargs: mutable.Set[List[Block]],
```

Comment:
captures are sets of names, not blocks. I would drop this here altogether and add type arguments instead.